### PR TITLE
Startup services logic on develop

### PR DESCRIPTION
**Pull request by sunil-indoria**

### CODE REVIEW

**Review by sunil-indoria — COMMENTED** [review]
Removed copy target as it is working without manually copying.
StartupServicesUtility has been added as a nuget reference in dependency.

File: nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj [hunk 1]
```diff
@@ -148,12 +148,18 @@
...
+        <TargetPath>build\netcoreapp2.0</TargetPath>
       </File>
     </ItemGroup>
     
     <ItemGroup>
       <Dependency Include="Microsoft.ServiceFabric.Services.Remoting">
         <Version>[$(NuGetPackageVersion)]</Version>
       </Dependency>
+      <Dependency Include="Microsoft.ServiceFabric.StartupServicesUtility">
+        <Version>[$(NuGetPackageVersion)]</Version>
```
Comment (sunil-indoria) [inline nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj:162]: Do I need to add this dependency in Microsoft.ServiceFabric.Actors.Wcf nuproj as well ?
I believe no as wfc has given dependency on Actors, hence it is covered.

File: src/Microsoft.ServiceFabric.Actors/Generator/ManifestGenerator.cs [hunk 2]
```diff
@@ -110,11 +111,139 @@ internal static void Generate(Arguments arguments)
...
+        }
+
+        private static StartupServicesManifestTypeParameter[] ConvertToServiceParameters(ApplicationManifestTypeParameter[] appParameters)
+        {
+            if (appParameters == null)
+            {
+                return null;
+            }
+
+            List<StartupServicesManifestTypeParameter> serviceParamsList = new List<StartupServicesManifestTypeParameter>();
```
Comment (sunil-indoria) [inline src/Microsoft.ServiceFabric.Actors/Generator/ManifestGenerator.cs:176]: Write appropriate logs for making it easier to debug

File: src/FabActUtil/ToolArguments.cs [hunk 3]
```diff
@@ -119,6 +119,27 @@ internal class ToolArguments
             LongName = "ServiceManifestEntryPointType",
             ShortName = "smep")]
         public string ServiceManifestEntryPointType;
+
+        [CommandLineArgument(
+            CommandLineArgumentType.AtMostOnce,
+            Description = "Absolute path of StartupServices file which contains DefaultServices.",
```
Comment (sunil-indoria) [inline src/FabActUtil/ToolArguments.cs:125]: and parameters

File: src/netstandard/FabActUtil/Microsoft.ServiceFabric.Actors.targets [hunk 4]
```diff
@@ -11,7 +11,13 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     <RuntimeVersion Include="$(_TargetFrameworkVersionWithoutV).0" />
     <FabActUtilRuntimeConfig Include="$(FabActUtilWorkingDir)\FabActUtil.runtimeconfig.json" />
   </ItemGroup>
-  <Target Name="_CopyAssembliesToFabActUtilWorkingDir">
+  <Target Name="_CopyAssembliesToTargetDir">
+    <ItemGroup>
```
[the remark this answers is not in the document]
Comment (sunil-indoria) [reply src/netstandard/FabActUtil/Microsoft.ServiceFabric.Actors.targets]: Checked. Not required

File: src/Microsoft.ServiceFabric.Actors/Generator/ManifestGenerator.cs [hunk 5]
```diff
@@ -110,11 +111,139 @@ internal static void Generate(Arguments arguments)
...
+                return null;
+            }
+
+            string defaultServicesXml = XmlSerializationUtility.Serialize(defaultServices);
+            XmlDocument defaultServicesDocument = new XmlDocument();
+            defaultServicesDocument.LoadXml(defaultServicesXml);
+
+            XmlDocument startupServicesDocument = new XmlDocument();
+            XmlElement startupDefaultServicesRoot = startupServicesDocument.CreateElement("StartupDefaultServicesType");
+            startupDefaultServicesRoot.InnerXml = defaultServicesDocument.DocumentElement.InnerXml;
```
[the remark this answers is not in the document]
Comment (sunil-indoria) [reply src/Microsoft.ServiceFabric.Actors/Generator/ManifestGenerator.cs:223]: at line number 216, defaultServicesXml has following values
<?xml version="1.0" encoding="utf-8"?>
<StartupDefaultServicesType xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <Service Name="Actor1ActorService" GeneratedIdRef="4521d3b2-29b8-434c-8fa0-6606f4f8c5ae|Persisted" xmlns="http://schemas.microsoft.com/2011/01/fabric">
    <StatefulService ServiceTypeName="Actor1ActorServiceType" TargetReplicaSetSize="[Actor1ActorService_TargetReplicaSetSize]" MinReplicaSetSize="[Actor1ActorService_MinReplicaSetSize]">
      <UniformInt64Partition PartitionCount="[Actor1ActorService_PartitionCount]" LowKey="-9223372036854775808" HighKey="9223372036854775807" />
    </StatefulService>
  </Service>
</StartupDefaultServicesType>

defaultServicesDocument.DocumentElement.InnerXml at line number 222 has this - 
<Service Name="Actor1ActorService" GeneratedIdRef="4521d3b2-29b8-434c-8fa0-6606f4f8c5ae|Persisted" xmlns="http://schemas.microsoft.com/2011/01/fabric"><StatefulService ServiceTypeName="Actor1ActorServiceType" TargetReplicaSetSize="[Actor1ActorService_TargetReplicaSetSize]" MinReplicaSetSize="[Actor1ActorService_MinReplicaSetSize]"><UniformInt64Partition PartitionCount="[Actor1ActorService_PartitionCount]" LowKey="-9223372036854775808" HighKey="9223372036854775807" /></StatefulService></Service>

InnerXml here has the required information without classtype so using this we can type cast from one class to another provided rest of the structure is same.

Here there is only

File: nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj [hunk 6]
```diff
@@ -148,12 +148,18 @@
       </File>
       <File Include="$(DropFolderNetStandard_Win)system.fabric.strings.dll">
         <TargetPath>build\netcoreapp2.0</TargetPath>
+      </File>
+	  <File Include="$(DropFolderNetStandard_Win)Microsoft.ServiceFabric.StartupServicesUtility.dll">
```
Comment (PavanKunapareddyMSFT) [inline nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj:152]: Do you need to include StartupServicesUtility.dll in DropFolderNetStandard path as well ?

File: src/FabActUtil/Tool.cs [hunk 7]
```diff
@@ -55,8 +55,20 @@ private static void GenerateOutput(ToolContext context)
             if (context.Arguments.Target == OutputTarget.Manifest)
             {
                 GenerateManifest(context);
-                AddParametersToLocalFiveNodeAppParamFile(context);
-                AddParametersToLocalOneNodeAppParamFile(context);
+
+                // If StartupServiceFilePath is provided, update parameters into service ParamFiles otherwise application ParamFiles
+                if (!string.IsNullOrEmpty(context.Arguments.StartupServicesFilePath))
+                {
+                    Console.WriteLine("Adding service parameters to service param files - {0} and {1}", context.Arguments.Local5NodeServiceParamFile, context.Arguments.Local1NodeServiceParamFile);
```
Comment (PavanKunapareddyMSFT) [inline src/FabActUtil/Tool.cs:62]: I assume Local5NodeServiceParamFile contains filename not the file content. If not, I am not sure if we should be tracing all the file content.

Comment (PavanKunapareddyMSFT) [reply src/FabActUtil/Tool.cs:62]: Got clarified by the below method. Its the filename.

File: src/Microsoft.ServiceFabric.Actors/Generator/ManifestGenerator.cs [hunk 8]
```diff
@@ -1425,6 +1597,16 @@ private string GetApplicationManifestFilePath()
                 return string.Empty;
             }
 
+            private string GetStartupServicesFilePath()
```
Comment (anantshankar17) [inline src/Microsoft.ServiceFabric.Actors/Generator/ManifestGenerator.cs:1600]: Minor: You could just replace separate method with this:

"return !string.IsNullOrEmpty(this.Arguments.StartupServicesFilePath) ? this.Arguments.StartupServicesFilePath : string.Empty;"

**Review by anantshankar17 — APPROVED** [review]

File: nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj [hunk 9]
```diff
@@ -94,7 +94,7 @@
       </File>
       <File Include="$(DropFolderNetFramework)\Microsoft.ServiceFabric.Actors.targets">
```
Comment (PavanKunapareddyMSFT) [inline nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj:95]: You need to bump the nuget version in props file as well inorder for Windows Fabric repo to consume your changes.

File: nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj (hunk 6 above)
Comment (sunil-indoria) [reply nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj:152]: In actor project when actor is build, it publish and put all required dlls at DropFolderNetStandard_Win 
From there this nuproj takes and put it under build\netcoreapp2.0

DropFolderNetStandard  is not required.

File: nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj (hunk 9 above)
Comment (sunil-indoria) [reply nuprojs/Microsoft.ServiceFabric.Actors/Microsoft.ServiceFabric.Actors.nuproj:95]: Bumped up to 5.0.4

**Review by PavanKunapareddyMSFT — APPROVED** [review]

**Pull request merged**